Require block arg type in signature

thomasmarshall opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->

This PR ensures explicit block arguments must be defined in the signature:

```rb
sig {void}
def foo(&blk); end
#        ^^^ error: Malformed `sig`. Type not specified for argument `blk`
```

The following will not produce an error:

```rb
sig {params(blk: T.proc.void).void}
def foo(&blk); end
```

Every method has a "synthetic" block argument if an explicit one isn't defined, for which an argument type **shouldn't** be required in the signature. We were previously determining if a block arg was synthetic or not based on its name reference. However, block arguments always have the name reference `<blk>` whether they are explicitly defined or synthetically added. This might not be intended, but is currently how it works:

https://github.com/sorbet/sorbet/blob/9adecfa7bffc4ffb6f72f806161709d6b97681af/namer/namer.cc#L801-L802

This PR swaps the name comparison for `isSyntheticBlockArgument`, which has a more robust way of determining if the block is synthetic or not, based on whether a `Loc` exists. Synthetic blocks use `Loc::none()`:

https://github.com/sorbet/sorbet/blob/9adecfa7bffc4ffb6f72f806161709d6b97681af/core/Symbols.cc#L2019-L2022

An alternative might be to change the namer so explicit block arguments are not named `<blk>`, but leaning on the `isSyntheticBlockArgument` seemed a smaller and simpler change.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

Fixes #5670. The motivation for the change is that all other arguments must have argument types if the signature is present, but block arguments are apparently ignored.

:wave: I picked this up as a "good first issue" as it seemed relatively straightforward to understand and limited in scope.

#### Notes

- The issue above mentions `typed: strict` but this change also applies in `typed: true` if a signature is defined, as it does for other argument types. Is this correct?
- The code was already in place for this to work, but the check for a "synthetic block" was apparently always returning true. Should we instead make that check work by changing `namer.cc`?
- I suspect this will produce errors in some codebases that previously didn't have them (as it did in some RBIs and tests here). Is that OK?

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

froydnj commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MiStEgJIoBidFj
→ https://go/builds/bui_MiStfRCE2frbmg

jez reviewed (approved):
Thanks for the change!

- - - - -

> * The issue above mentions `typed: strict` but this change also applies in `typed: true` if a signature is defined, as it does for other argument types. Is this correct?

```ruby
# typed: true
extend T::Sig

sig {void}
def example
  yield
end

example do
  puts('hello')
end
```

This is not an error in `# typed: true` but is an error in `# typed: strict`, even on your branch. At the moment, this is intended for ease of migration on legacy codebases, but we could reconsider this behavior in the future. (This would be good to add a test for, btw.)

> * I suspect this will produce errors in some codebases that previously didn't have them (as it did in some RBIs and tests here). Is that OK?

That depends on how many errors, and how easy it is to fix the problem.

For the question of "how many," it looks like Stripe's codebase of multiple millions of lines of code only has 6 new errors because of this change, so I'm inclined to say that we're good to go here.

If you wanted to make it even more palatable to roll this change out, you could consider adding an autocorrect for this error message, but I think that's both tricky (going to take some time to implement) and also likely to be rare (as evidenced by the 6 errors above), so I wouldn't block landing the change on it.

> * The code was already in place for this to work, but the check for a "synthetic block" was apparently always returning true. Should we instead make that check work by changing namer.cc?

I'm sorry, I don't know what you mean.

jez reviewed (changes requested):
I notice that in certain situations, the error message is less than idea:

```ruby
# typed: true
class Example
  extend T::Sig
  sig {params(args: T.untyped).void}
  def foo(args); end
end
class Example
  def foo(args, &my_custom_block_name); end
end
```

The error message here:

```
❯ sorbet foo.rb
foo.rb:8: Malformed sig. Type not specified for argument <blk> https://srb.help/5003
     8 |  def foo(args, &my_custom_block_name); end
                         ^^^^^^^^^^^^^^^^^^^^
    foo.rb:4: Signature
     4 |  sig {params(args: T.untyped).void}
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 1
```

Do you think you could tweak the code to use the name that comes from the method def instead of from the symbol, so that the error message always matches what the user wrote?

thomasmarshall commented:
> This is not an error in `# typed: true` but is an error in `# typed: strict`, even on your branch.

Ah OK, but the example from #5670 _is_ now an error for both `# typed: true` and `# typed: strict`. I think that's correct, given it's the same behaviour as other arguments, but I just wanted to double check my understanding of the issue. I've added a test for unnamed block/yield.

> I'm sorry, I don't know what you mean.

I mean the `bool isSyntheticBlkArg` variable already existed but the comparison `arg.name == core::Names::blkArg()` was incorrect. I thought maybe that comparison is _supposed_ to be correct, and that we should be able to identify a synthetic block by inspecting the arg name. I was speculating that the bug might be in the thing that sets the argument name, rather than where I've made the change. But I'm happy that `isSyntheticBlockArgument` works well!

> Do you think you could tweak the code to use the name that comes from the method def instead of from the symbol, so that the error message always matches what the user wrote?

I've added a fixup and test for this now:

```
foo.rb:8: Malformed sig. Type not specified for argument my_custom_block_name https://srb.help/5003
     8 |  def foo(args, &my_custom_block_name); end
                         ^^^^^^^^^^^^^^^^^^^^
    foo.rb:4: Signature
     4 |  sig {params(args: T.untyped).void}
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 1
```
